Accept Shop Data when Inviting Shop Owner

pmn4 opened the pull request:
Feature Request: Accept Shop Data when Inviting a Shop Owner (not just _User_ data)

This PR was prompted by a Shop Importer that I am working on.
I would like to leverage the `inviteShopOwner` code which creates an account, a new shop, kicks off emails, etc., but currently creates a shop with a copy of the Primary Shop (e.g., the name is "Reaction Shop 1").

As far as testing goes, there should be no impact on current functionality since there are no UI changes. Server code can now call `inviteShopOwner` with both user data _and_ shop data: `Meteor.call("accounts/inviteShopOwner", userData, shopData)`.
There is a test to show how it might be used.

pmn4 commented on server/methods/accounts/accounts.app-test.js:
```diff
@@ -551,7 +553,7 @@ describe("Account Meteor method ", function () {
           name: fakeUser.profile.addressBook[0].fullName
         })
-      ).to.not.throw(Meteor.Error, /Access denied/);
+      ).to.not.throw(Meteor.Error);
```
this is not necessarily related to this change, but I coded an error by mistake and this test was still passing.
I find it good practice to not limit the _type_ of error thrown in a ".to.not.throw" test, so I changed it. lmk if you'd like to revert

pmn4 commented on server/methods/accounts/accounts.app-test.js:
```diff
@@ -563,9 +565,32 @@ describe("Account Meteor method ", function () {
+      const primaryShop = getShop();
+      const shopData = {
+        name: shopName,
+        locales: primaryShop.locales
```
_(unfortunately `locales` is required and does not have a defaultValue)_

pmn4 commented on server/methods/accounts/accounts.js:
```diff
@@ -592,7 +594,7 @@ export function inviteShopOwner(options) {
   }
 
-  const { shopId } = Meteor.call("shop/createShop", userId) || {};
+  const { shopId } = Meteor.call("shop/createShop", userId, shopData) || {};
```
`shop/createShop` already accepts shop data, so the major change here is accepting a new parameter and passing it on to the createShop method

pmn4 commented on server/methods/accounts/accounts.js:
```diff
@@ -912,10 +914,19 @@ export function setUserPermissions(userId, permissions, group) {
 function getEmailLogo(shop) {
   let emailLogo;
   if (Array.isArray(shop.brandAssets)) {
+    let mediaId;
```
the changes from here on add extra null-checking/default value setting.
I changed them to make tests less verbose (which is typically is a bad reason for changing code!), but also figured extra null checks are never a bad thing.

pmn4 commented on server/methods/core/shop.js:
```diff
@@ -81,17 +80,24 @@ Meteor.methods({
+      const count = Collections.Shops.find().count() || "";
+      seedShop = Reaction.getPrimaryShop();
+      // generate unique shop name
+      seedShop.name = seedShop.name + count;
```
Previously, `shopData.name` would get a number appended regardless of whether the shopData is a clone of PrimaryShop or passed by the calling function. Now, this only happens when Primary Shop is cloned.

pmn4 commented on server/methods/core/shop.js:
```diff
@@ -121,16 +127,16 @@ Meteor.methods({
 
     // we should have created new shop, or errored
-    Logger.info("Created shop: ", shop._id);
+    Logger.info("Created shop: ", newShopId);
```
better to use the _id returned from Shops.insert just in case the passed value is ignored (and auto generated... which... it really should be, right?)

pmn4 commented on server/methods/accounts/accounts.js:
```diff
@@ -910,15 +912,21 @@ export function setUserPermissions(userId, permissions, group) {
-    emailLogo = path.join(Meteor.absoluteUrl(), mediaId.url());
-  } else {
-    emailLogo = Meteor.absoluteUrl() + "resources/email-templates/shop-logo.png";
+    let media;
```
the changes from here add extra null-checking.
I changed them to make tests a little easier to write (which is typically not a great reason for changing code, I know!), but also figured extra null checks are never a bad thing.

pmn4 commented:
odd... tests pass when I run only the account.app-test.js file

pmn4 commented on server/startup/accounts.js:
```diff
@@ -120,9 +120,14 @@ export default function () {
+          additionals.groups = [group._id];
+        }
         // if no group or customer permissions retrieved from DB, use the default Reaction customer set
-        roles[shopId] = group.permissions || Reaction.defaultCustomerRoles;
```
new test was failing on this line, where `group` was null.
While I don't understand how similar tests were passing, an extra null check can't hurt.

zenweasel commented:
I'm getting an error when create a shop through the dashboard now

```
Exception while invoking method 'accounts/inviteShopOwner' Error: Match error: Failed Match.OneOf, Match.Maybe or Match.Optional validation
    at exports.check (packages/check.js:55:15)
    at DDPCommon.MethodInvocation.shopCreateShop (server/methods/core/shop.js:30:5)
    at packages/check.js:128:16
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:1686:15)
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at resolve (packages/ddp-server/livedata_server.js:1684:36)
    at new Promise (<anonymous>)
    at Server.applyAsync (packages/ddp-server/livedata_server.js:1683:12)
    at Server.apply (packages/ddp-server/livedata_server.js:1622:26)
    at Server.call (packages/ddp-server/livedata_server.js:1604:17)
    at DDPCommon.MethodInvocation._inviteShopOwner (server/methods/accounts/accounts.js:597:29)
    at packages/check.js:128:16
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at Object._failIfArgumentsAreNotAllChecked (packages/check.js:127:41)
    at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1765:18)
    at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:719:19)
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
    at DDPServer._CurrentWriteFence.withValue (packages/ddp-server/livedata_server.js:717:46)
    at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1186:15)
 => awaited here:
```

pmn4 commented on imports/plugins/included/marketplace/client/components/inviteOwner.js:
```diff
@@ -31,8 +31,9 @@ class InviteOwner extends Component {
+    const shopData = {}; // TODO: add optional shop data to the form (maybe just name?)
 
-    Meteor.call("accounts/inviteShopOwner", { name, email }, (error, result) => {
+    Meteor.call("accounts/inviteShopOwner", { name, email }, shopData, (error, result) => {
```
oops! passing `{}` fails the `Match.Optional(Schemas.Shop)` check.

pmn4 commented:
@aldeed does the new Simple Schema provider richer error messages when things like `Match.OneOf`, `Match.Maybe`, or `Match.Optional` fail validation? The messages when you check against a schema directly are super useful. I've had to change the validation when these fail to find out what's really going on, then change them back. Not the end of the world, but would prefer not to have to do that!

pmn4 commented on server/methods/accounts/accounts.app-test.js:
```diff
@@ -66,7 +66,7 @@ describe("Account Meteor method ", function () {
     });
 
-    it("should allow user to add new addresses", function (done) {
+    it("should allow user to add new addresses", function () {
```
I removed all of the `done` callbacks based on @zenweasel's comment (none of these tests are async)

pmn4 commented on server/methods/accounts/accounts.js:
```diff
@@ -910,15 +903,21 @@ export function setUserPermissions(userId, permissions, group) {
-    emailLogo = Meteor.absoluteUrl() + "resources/email-templates/shop-logo.png";
+    let media;
+    const brandAsset =
+      shop.brandAssets.find((asset) => asset.type === "navbarBrandImage");
```
s/`_.find`/`Array#find`

pmn4 commented on server/methods/accounts/accounts.js:
```diff
@@ -741,16 +743,7 @@ export function sendWelcomeEmail(shopId, userId) {
-  // Get shop logo, if available. If not, use default logo from file-system
-  let emailLogo;
-  if (Array.isArray(shop.brandAssets)) {
-    const brandAsset = _.find(shop.brandAssets, (asset) => asset.type === "navbarBrandImage");
```
started to replace this `_.find` and realized there is a method which already does this

zenweasel commented:
@pmn4 Hey Pat, is this ready for re-review

pmn4 commented:
@zenweasel it is, thanks!
do you have a process for re-reviewing PRs, or should I @ you when I am done making changes?

aldeed commented on server/methods/core/shop.js:
```diff
@@ -81,17 +80,24 @@ Meteor.methods({
+
+    // ensure unique id
+    // (shouldn't Mongo handle this for us? with lesser risk of collision?)
+    seedShop._id = Random.id();
```
Note that I have some significant changes to this in https://github.com/reactioncommerce/reaction/pull/3331, which most likely will be merged before this PR and require some conflict resolution. My changes include letting Mongo/Meteor generate the `_id` as you mention.

https://github.com/reactioncommerce/reaction/pull/3331/files#diff-a85b4a331f34703f51a72e9728869fe2

pmn4 commented on server/methods/accounts/accounts.app-test.js:
```diff
@@ -528,8 +516,7 @@ describe("Account Meteor method ", function () {
           name: fakeUser.profile.addressBook[0].fullName
-        })).to.not.throw(Meteor.Error, /Access denied/);
-      return done();
+        })).to.not.throw(Meteor.Error);
```
when testing for errors not thrown, it is a good idea to remove the limitation on _type_ of error.

pmn4 commented:
@zenweasel fixed a merge conflict, but otherwise, nothing has changed in the past few days since I addressed your comments. let me know if you have any other questions.

aaronjudd commented:
@reactioncommerce/community can we move this PR forward for 1.9, shepherds needed..

zenweasel commented:
@pmn4 Hey Pat, thanks for your patience with this. If you can fix up these conflicts (I tried but they didn't seem obvious to me) I will make sure and get this review and hopefully approved today.

zenweasel commented:
@pmn4 If we could get these conflicts resolved that would be great. I tried again today and there's a couple around `brandAssets` I couldn't make the call. If we could get those resolved we should be able to move forward quickly.

zenweasel commented:
@pmn4 Sorry Pat, we need to get the conflicts resolved again. i tried (again) and it got too dodgy.

pmn4 commented:
haha, no problem. I will take a look tomorrow

zenweasel commented:
@pmn4 We still have a few conflicts here

zenweasel commented:
@pmn4 Can we get the conflicts on this resolved? I don't see any other thing that's holding us up.

zenweasel commented:
Actually, looks like we also have failing tests

pmn4 commented:
It's been a few days since I was able to take a look at this @zenweasel, but should be green now

zenweasel commented:
When I invite a new shop owner I get this empty error message and no shop is created.

![2018-04-11_07-27-31](https://user-images.githubusercontent.com/72819/38588705-dc2850de-3d59-11e8-9728-483c04ac40cd.png)

pmn4 commented:
ah! sorry for wasting your time @zenweasel. I had been testing on an instance with a "cleaner" primary shop. the `"taxes"` in Shops.json violates the Shop schema. I will address that in a separate PR.
